### PR TITLE
RichText: own undo signalling

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

This is a proposal for the `RichText` component to handle undo signalling in favour of listening to the TinyMCE `change` event.

`RichText` will signal that an undo level should be created:

* When the user changes the value, except when it comes from the `input` event (typing).
* When there is over a second pause between input events (typing).

Why?

It's not always clear when the `change` event exactly triggers and we do not always want it to create an undo level. The `change` event also no longer accounts for formatting changes, as these are no longer done by TinyMCE. It's easier to just do it ourselves so we can fine-tune and not rely on the TinyMCE event.

## How has this been tested?
Test undoing and redoing changes from `RichText` components.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

**Comment by iseulde** [comment]
TinyMCE's logic for undo levels, for reference:

https://github.com/tinymce/tinymce/blob/dd51681f898eea7e61fa9b4c2ef849aaf6853920/src/core/main/ts/api/UndoManager.ts#L72-L173

**Comment by iseulde** [comment]
@aduth @mcsf This branch now has the undo timeout consistently, but I cannot figure out what exactly the problem is. I can also reproduce it locally in headless mode, but as soon as I turn headless off, it does not occur at all, even at slowo=0. It seems there's only one extra undo level being created, if I undo once more in the test it passes. I've not been able to figure it out further.

**Comment by aduth** [comment]
>Moves the mouse or touches the screen.

Why this? I don't like the idea of every single RichText binding to `mousemove`. That's easily thousands of callbacks in a single cursor gesture.

**Comment by iseulde** [comment]
> Why this? I don't like the idea of every single RichText binding to mousemove. That's easily thousands of callbacks in a single cursor gesture.

Yeah that's true. We want to create undo levels as soon as typing is interrupted. Here listening for mousedown instead will probably suffice.

File: packages/editor/src/utils/with-history/index.js [hunk 1]
```diff
@@ -112,6 +112,12 @@ const withHistory = ( options = {} ) => ( reducer ) => {
 			return state;
 		}
 
+		// Temporary: illustrates the problem with RECEIVE_BLOCKS creating an
+		// extra undo level at a random time.
+		if ( action.type === 'RECEIVE_BLOCKS' ) {
```
Comment (iseulde) [inline packages/editor/src/utils/with-history/index.js]: @aduth This fixes the "Should undo to expected level intervals" e2e test for me. The problem seems to be that `RECEIVE_BLOCKS` creates an extra undo level at a random time, when typing happens fast on first page load. I'm not quite sure what this action does in the first place. If this is the point at which we receive content from the server, it seems bad to allow content modification before this action occurs.

Comment (iseulde) [reply packages/editor/src/utils/with-history/index.js]: > Unlike resetBlocks, these should be appended to the existing known set, not replacing.

Huh. What did the user do to cause this?

@noisysocks The only use of this action seems to be reusable blocks. Any idea why its being dispatched when the editor loads, and why it would create an undo level for it?

Comment (iseulde) [reply packages/editor/src/utils/with-history/index.js]: Never mind, resolved it in #11399. :)

File: packages/editor/src/components/rich-text/index.js [hunk 2]
```diff
@@ -914,6 +883,7 @@ export class RichText extends Component {
 						</Fragment>
 					) }
 				</Autocomplete>
+				<RemoveBrowserShortcuts />
```
Comment (aduth) [inline packages/editor/src/components/rich-text/index.js]: Another instance of: Should every `RichText` be creating these global event bindings? Worse yet, wouldn't this also disable handling of those shortcuts when not even within a `RichText`, e.g. the title input?

File: packages/editor/src/components/rich-text/index.js [hunk 3]
```diff
@@ -113,14 +111,19 @@ export class RichText extends Component {
 		this.state = {};
 
 		this.usedDeprecatedChildrenSource = Array.isArray( value );
+		this.lastHistoryValue = value;
 	}
 
 	componentDidMount() {
 		document.addEventListener( 'selectionchange', this.onSelectionChange );
+		window.addEventListener( 'mousedown', this.onCreateUndoLevel );
```
Comment (aduth) [inline packages/editor/src/components/rich-text/index.js]: Still. Why are we binding regardless of whether the `RichText` is selected?

Comment (aduth) [reply packages/editor/src/components/rich-text/index.js]: I still don't understand what a mouse gesture has to do with undo, nor agree with assumption that use of mouse implies some milestone in the user's writing flow. Was this inherited from TinyMCE?

Comment (aduth) [reply packages/editor/src/components/rich-text/index.js]: As written, this would be a good candidate for `withGlobalEvents`:

https://github.com/WordPress/gutenberg/tree/master/packages/compose/src/with-global-events

Comment (iseulde) [reply packages/editor/src/components/rich-text/index.js]: Oh, I see. Sure we can bind on selection, unbind on deselect.

> I still don't understand what a mouse gesture has to do with undo, nor agree with assumption that use of mouse implies some milestone in the user's writing flow. Was this inherited from TinyMCE?

Yes, TinyMCE has the same logic, which I think makes sense. E.g. if you type something, then move to another place in the text and continue to write, the undo button should not undo all typing at once.

File: packages/editor/src/components/rich-text/index.js (hunk 2 above)
Comment (iseulde) [reply packages/editor/src/components/rich-text/index.js]: You're right. Maybe the place to put these would be more up the tree in the Editor component to disable it for all fields. No reason to bind multiple times. In the case af the title, disabling seems wanted as well, otherwise undoing with the browser will add levels in Gutenberg.

File: packages/editor/src/components/rich-text/index.js [hunk 4]
```diff
@@ -780,6 +737,18 @@ export class RichText extends Component {
 			record.end = length;
 			this.applyRecord( record );
 		}
+
+		if ( isSelected && ! prevProps.isSelected ) {
```
Comment (iseulde) [inline packages/editor/src/components/rich-text/index.js]: @aduth Better?

Comment (aduth) [reply packages/editor/src/components/rich-text/index.js]: We still have a condition in `onSelectionChange` which tests whether the rich text is the active one, which sends mixed messaging as to whether the intent of the application of event handlers _here_ is meant to ensure they're only bound if the field is active/selected.

File: packages/editor/src/components/rich-text/remove-browser-shortcuts.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
+/**
+ * WordPress dependencies
+ */
+import { rawShortcut } from '@wordpress/keycodes';
+import { KeyboardShortcuts } from '@wordpress/components';
+
+export function RemoveBrowserShortcuts() {
```
Comment (aduth) [inline packages/editor/src/components/rich-text/remove-browser-shortcuts.js]: Is this removing from TinyMCE as well? How do _our_ handlers for Cmd+Z Undo not get removed as well?

The fact it works as-is all seems rather coincidental, not something I'd think to rely on.

File: test/e2e/specs/undo.test.js [hunk 6]
```diff
@@ -10,10 +10,52 @@ import {
...
+
+		await page.keyboard.type( 'before move' );
+		await page.mouse.down();
+		await page.keyboard.type( ' after move' );
+
+		expect( await getEditedPostContent() ).toMatchSnapshot();
+
+		await pressWithModifier( META_KEY, 'z' );
+
+		expect( await getEditedPostContent() ).toMatchSnapshot();
```
Comment (aduth) [inline test/e2e/specs/undo.test.js]: We should ideally test caret position here too. Because it resets to the beginning of the field unexpectedly. Apparently this also occurs in the latest plugin release, so not strictly a regression.

File: test/e2e/specs/blocks/__snapshots__/quote.test.js.snap [hunk 7]
```diff
@@ -58,11 +58,7 @@ exports[`Quote can be converted to paragraphs and renders a paragraph for the ci
 <!-- /wp:paragraph -->"
 `;
 
-exports[`Quote can be converted to paragraphs and renders a void paragraph if both the cite and quote are void 1`] = `
-"<!-- wp:paragraph -->
-<p></p>
-<!-- /wp:paragraph -->"
-`;
+exports[`Quote can be converted to paragraphs and renders a void paragraph if both the cite and quote are void 1`] = `""`;
```
Comment (aduth) [inline test/e2e/specs/blocks/__snapshots__/quote.test.js.snap]: I saw this change in another pull request. What's going on with this? It doesn't seem like we should be expecting any changes here.

File: test/e2e/specs/__snapshots__/undo.test.js.snap [hunk 8]
```diff
@@ -13,3 +13,39 @@ exports[`undo Should undo to expected level intervals 1`] = `
 <p>test</p>
 <!-- /wp:paragraph -->"
 `;
+
+exports[`undo should undo typing after arrow navigation 1`] = `
+"<!-- wp:paragraph -->
+<p>before keyboar after keyboardd</p>
```
Comment (aduth) [inline test/e2e/specs/__snapshots__/undo.test.js.snap]: The result here _looks_ off, but I guess the test was written in a way where it's expected. Might catch a future maintainer off-guard.

File: packages/editor/src/components/rich-text/remove-browser-shortcuts.js (hunk 5 above)
Comment (aduth) [reply packages/editor/src/components/rich-text/remove-browser-shortcuts.js]: >Is this removing from TinyMCE as well?

Oh, I see these are removed in `TinyMCE`'s initialization.

File: packages/editor/src/components/rich-text/tinymce.js [hunk 9]
```diff
@@ -209,13 +209,15 @@ export default class TinyMCE extends Component {
 
 				editor.on( 'init', () => {
 					// See https://github.com/tinymce/tinymce/blob/master/src/core/main/ts/keyboard/FormatShortcuts.ts
-					[ 'b', 'i', 'u' ].forEach( ( character ) => {
+					[ 'b', 'i', 'u', 'z', 'y', 'z' ].forEach( ( character ) => {
```
Comment (aduth) [inline packages/editor/src/components/rich-text/tinymce.js]: `z` twice.

File: packages/editor/src/components/rich-text/remove-browser-shortcuts.js [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+	return (
+		<KeyboardShortcuts
+			bindGlobal
+			shortcuts={ {
+				[ rawShortcut.primary( 'z' ) ]: () => false,
```
Comment (aduth) [inline packages/editor/src/components/rich-text/remove-browser-shortcuts.js]: This feels a bit like we're relying a bit too much on incidental support of the `false` return offered as a "convenience" by Mousetrap, but otherwise not the canonical way for preventing default behavior:

https://craig.is/killing/mice#api.bind.default

**Comment by aduth** [comment]
The Undo behavior feels particularly heavy-handed (destroying all of a paragraph I've written if I don't move my mouse), but I guess this aligns with the previous behavior as well.

File: packages/editor/src/components/rich-text/index.js (hunk 4 above)
Comment (aduth) [reply packages/editor/src/components/rich-text/index.js]: Furthermore, it's unclear what distinction is meant to be made between `this.isActive()` and `this.props.isSelected`, which seems important to clarify as we're making assumptions about `isSelected` as a basis for assuming an active RichText.

Comment (aduth) [reply packages/editor/src/components/rich-text/index.js]: Looking through how `isSelected` becomes assigned, I don't know that it's safe currently to make the assumption that `isSelected` can imply the field is focused. We only call `setFocusedElement` on `focus`, but never anything when focus leaves the field. While only one RichText can be the focused element at a given time, I still think it's possible for a `RichText` to be the "focused instance" despite the _actual DOM focus_ being elsewhere in the page (title, for example). This feels like it should be improved, and blocks being able to use `isSelected` here.

Ideally, I think we consolidate `isActive` with `isSelected` such that `isSelected` can only be true if and only if the `RichText` field is focused.

File: packages/editor/src/components/rich-text/tinymce.js (hunk 9 above)
Comment (aduth) [reply packages/editor/src/components/rich-text/tinymce.js]: > `z` twice.

Corrected in e69e6ff.

File: packages/editor/src/components/rich-text/remove-browser-shortcuts.js (hunk 5 above)
Comment (aduth) [reply packages/editor/src/components/rich-text/remove-browser-shortcuts.js]: > > Is this removing from TinyMCE as well?
> 
> Oh, I see these are removed in `TinyMCE`'s initialization.

Clarified some in e69e6ff.

File: packages/editor/src/components/rich-text/remove-browser-shortcuts.js (hunk 10 above)
Comment (aduth) [reply packages/editor/src/components/rich-text/remove-browser-shortcuts.js]: > This feels a bit like we're relying a bit too much on incidental support of the `false` return offered as a "convenience" by Mousetrap, but otherwise not the canonical way for preventing default behavior:
> 
> https://craig.is/killing/mice#api.bind.default

Improved in 41ed725.

**Comment by iseulde** [comment]
> The Undo behavior feels particularly heavy-handed (destroying all of a paragraph I've written if I don't move my mouse), but I guess this aligns with the previous behavior as well.

It should behave roughly the same as before. With this change we can also fine tune it ourselves. Ideas for other triggers are welcome. Not sure what else we can do, maybe a typing pause could trigger an undo level? I don't think we can rely on punctuation as that's not universal.

File: packages/editor/src/components/rich-text/remove-browser-shortcuts.js (hunk 10 above)
Comment (iseulde) [reply packages/editor/src/components/rich-text/remove-browser-shortcuts.js]: Thanks!

File: packages/editor/src/components/rich-text/tinymce.js (hunk 9 above)
Comment (iseulde) [reply packages/editor/src/components/rich-text/tinymce.js]: Thanks!

File: test/e2e/specs/__snapshots__/undo.test.js.snap (hunk 8 above)
Comment (iseulde) [reply test/e2e/specs/__snapshots__/undo.test.js.snap]: I'll investigate. This seems wrong.

File: packages/editor/src/components/rich-text/index.js (hunk 4 above)
Comment (gziolo) [reply packages/editor/src/components/rich-text/index.js]: I think [the comment](https://github.com/WordPress/gutenberg/pull/6871#issuecomment-392895193) from @youknowriad is the best way to describe why `blur` is very hard to implement in this case:

> The issue is that sometimes RichText loses the focus but we still want to keep it selected. (Link modal, toolbar buttons), It's probably doable but not simple because of the fact that popovers show outside the RichText's node.

This issue has a very long history. Initially focus had to be manually maintained using `isSelected` and `onFocus` prop. `onFocus` was proxied through `RichText` to `TinyMCE` but it has changed in #6286. I tried to get rid of their usage in core with #6871 but we decided that it isn't ideal because it doesn't handle `blur` event at all. To move forward we opened #7029 which renamed `onFocus` to `unstableOnFocus`.

> Ideally, I think we consolidate `isActive` with `isSelected` such that `isSelected` can only be true if and only if the `RichText` field is focused.

I'm not quite sure what role `isActive` plays now. I don't think it existed when `setFocusedElement` was introduced. 

```js
isActive() {
	return this.editableRef === document.activeElement;
}
```
Okey, so it seems like it has the same limitations as working with `blur` event. It might not work as expected when using popovers or block toolbar controls.

We need to find a good solution because this also causes some confusing bugs in the UI when building custom blocks which contain other components which can take over focus like `PlainText` of any form controls. See #7463 for reference.

Comment (aduth) [reply packages/editor/src/components/rich-text/index.js]: This might be something we could use [`withFocusOutside`](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/higher-order/with-focus-outside) for, which handles both cases of:

- `blur` is actually a focus change _within_ the element
- _within_ the element might not _within_ the same DOM tree (`withFocusOutside` uses React's virtual bubbling)

Comment (gziolo) [reply packages/editor/src/components/rich-text/index.js]: It works perfectly fine with popover integration for `Link` format control. However it doesn't work properly with mouse interaction in the block toolbar because this slot is rendered outside of the `RichText` component.

**Comment by iseulde** [comment]
@aduth I went for adding an undo level after a one second pause in input. We can adjust this value if it's too high or low.

File: packages/editor/src/components/rich-text/index.js [hunk 11]
```diff
@@ -612,7 +573,7 @@ export class RichText extends Component {
 		// The input event does not fire when the whole field is selected and
 		// BACKSPACE is pressed.
 		if ( keyCode === BACKSPACE ) {
-			this.onChange( this.createRecord(), true );
```
Comment (aduth) [inline packages/editor/src/components/rich-text/index.js]: I don't understand why this change has been made.

File: test/e2e/specs/undo.test.js [hunk 12]
```diff
@@ -10,10 +10,38 @@ import {
...
-	beforeAll( async () => {
+	beforeEach( async () => {
 		await newPost();
 	} );
 
+	it( 'should undo typing after a pause', async () => {
+		await clickBlockAppender();
+
+		await page.keyboard.type( 'before pause' );
+		await new Promise( ( resolve ) => setTimeout( resolve, 1000 ) );
```
Comment (aduth) [inline test/e2e/specs/undo.test.js]: I wish we could emulate this without _actually_ slowing our test run by one full second.

File: packages/editor/src/components/rich-text/index.js (hunk 11 above)
Comment (iseulde) [reply packages/editor/src/components/rich-text/index.js]: We used to skip applying changes back to the live DOM, and so did we here. Since #11595 we no longer do that, but it hasn't been updated here. I've removed it here while revising the parameters for `this.onChange`.

File: packages/editor/src/components/editor-global-keyboard-shortcuts/index.js [hunk 13]
```diff
@@ -114,6 +99,9 @@ class EditorGlobalKeyboardShortcuts extends Component {
 					bindGlobal
 					shortcuts={ {
 						[ rawShortcut.primary( 's' ) ]: this.save,
+						[ rawShortcut.primary( 'z' ) ]: flow( preventDefault, onUndo ),
```
Comment (aduth) [inline packages/editor/src/components/editor-global-keyboard-shortcuts/index.js]: By moving this into `bindGlobal`, we're now preventing default behavior for _all_ inputs in the page, not just RichText, right? We'd then rely on all of those inputs to be fully controlled by state values governed by history? I'm not sure I feel particularly confident we can trust that.

Comment (iseulde) [reply packages/editor/src/components/editor-global-keyboard-shortcuts/index.js]: Why is the title not part of the history?

Comment (aduth) [reply packages/editor/src/components/editor-global-keyboard-shortcuts/index.js]: Non-exhaustive set of others:

- Embed placeholder URL input
- Table placeholder columns / rows inputs
- Media placeholders URL input
- Post permalink
- Categories search

(Discovered via `onSubmit=` Find in Files search)

Comment (iseulde) [reply packages/editor/src/components/editor-global-keyboard-shortcuts/index.js]: Okay, I'm unsure atm then how to prevent default behaviour for RichText only. Restore https://github.com/WordPress/gutenberg/blob/72e51c692108ce52021564352a6d2661466fc231/packages/editor/src/components/rich-text/remove-browser-shortcuts.js?

Comment (aduth) [reply packages/editor/src/components/editor-global-keyboard-shortcuts/index.js]: Yeah, for now I think we might just need to have two KeyboardShortcuts for handling these keys, one top-level but not applied to inputs, the other specific to when a RichText is selected.

tl;dr I think that's what we had in the prior revision you're referring to?

Comment (iseulde) [reply packages/editor/src/components/editor-global-keyboard-shortcuts/index.js]: Right, I just restored your revision :)

File: test/e2e/specs/blocks/__snapshots__/quote.test.js.snap (hunk 7 above)
Comment (iseulde) [reply test/e2e/specs/blocks/__snapshots__/quote.test.js.snap]: See https://github.com/WordPress/gutenberg/pull/11595#discussion_r232256521.

**Review by aduth — APPROVED** [review]

File: packages/editor/src/components/rich-text/remove-browser-shortcuts.js [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+/**
+ * Set of keyboard shortcuts handled internally by RichText.
+ *
+ * @type {Array}
+ */
+const HANDLED_SHORTCUTS = [
+	rawShortcut.primary( 'z' ),
+	rawShortcut.primaryShift( 'z' ),
+	rawShortcut.primary( 'y' ),
```
[the remark this answers is not in the document]
Comment (iseulde) [reply packages/editor/src/components/rich-text/remove-browser-shortcuts.js]: We're only removing the browser shortcut here, not adding any support. But maybe we should add support everywhere?

**Comment by iseulde** [comment]
Thanks a lot @aduth for the review!

File: packages/editor/src/components/rich-text/remove-browser-shortcuts.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (aduth) [reply packages/editor/src/components/rich-text/remove-browser-shortcuts.js]: Hmm, in that case, is the line even necessary if we're not handling it anyways?
